Some updates to make the debian packageing work.

Kunzol opened the pull request:
## Updates to make debuild work again.

Because the directory was moved to the subdir dist, it is needed
to copy the directory to the top directory or set a link
before running debuild.

As usual the command for building the debian package is:
```
debuild -b -us -uc
```
To make it configurable some environment variables can be set.
Due to the fact, that debuild filters out some variables, all the
configuration variables need to be prefixed by DEB_

First the changelog file is updated with the current version. 
To configure the fullname and email available in the changelog
message set the environment variables:
```
export DEBFULLNAME="Your Name"
export DEB_FULLNAME=${DEBFULLNAME}
export EMAIL=<your email>
```
options for mconfig, by setting the environment variables:
`DEB_NOSUID=1 `   adds --without-suid
`DEB_NONETWORK=1` adds --without-network
`DEB_NOSECCOMP=1` adds --without-seccomp
`DEB_NOALL=1`     adds all of the above

To select a specific profile for mconfig:
```
export DEB_SC_PROFILE=release-stripped
```
or
```
export DEB_SC_PROFILE=debug
```
The builddir can be specified with this:
```
export DEB_SC_BUILDDIR=builddir
```
Currently no other config options can be set via environment,
but it would be easy to add them if needed.

I used as much as possible from the existing rules file for this
first step.

In the control file I added a build dependency to golang and changed
the architecture to amd64.

I do not expect that it is already perfect, but I hope this helps others in need 
for a debian package of singularity-container.

Kunzol committed:
Some updates to make the debian packageing work.

Because the directory was moved to the subdir dist, it is needed
to copy the directory to the top directory or set a link
before running debuild.

As usual the command for building the debian package is:
debuild -b -us -uc

To make it configurable some environment variables can be set.
Due to the fact, that debuild filters out some variables, all the
configuration variables need to be prefixed by DEB_

options for mconfig, by setting the environment variables:
DEB_NOSUID=1    adds --without-suid
DEB_NONETWORK=1 adds --without-network
DEB_NOSECCOMP=1 adds --without-seccomp
DEB_NOALL=1     adds all of the above

To select a specific profile for mconfig:
DEB_SC_PROFILE=release-stripped
DEB_SC_PROFILE=debug

The builddir can be specified with this:
DEB_SC_BUILDDIR=builddir

Currently no other config options can be set via environment,
but it would be easy to add them if needed.

I used as much as possible from the existing rules file for this
first step.

In the control file I added a build dependency to golang and changed
the architecture to amd64.

Kunzol committed:
added detailed info about the build process.

DrDaveD reviewed (changes requested):
Thanks for the description file.  Please add a reference to it in INSTALL.md too so people can find it.  I think a new short section at the end is appropriate.

Kunzol committed:
subsection for debian added.

Kunzol committed:
corrected some spelling errors.

DrDaveD commented on INSTALL.md:
```diff
@@ -161,3 +161,9 @@ $ make -C builddir rpm RPMPREFIX=/usr/local
+## Debian Package
+
+Additional information on [how to build a Debian package](dist/debian/DEBIAN_PACKAGE.md) can be 
+found in `dist/debian`.
```
```suggestion
Additional information on how to build a Debian package can be found in [dist/debian/DEBIAN_PACKAGE.md](dist/debian/DEBIAN_PACKAGE.md).
```

DrDaveD commented on dist/debian/control:
```diff
@@ -15,14 +15,15 @@ Build-Depends:
 # "singularity" is a packaged game (but the contents don't clash)
 Package: singularity-container
-Architecture: any
+Architecture: amd64
```
What happens if you don't make this change?  I still have an i386 debian machine and it complains about this.

Kunzol replied:
I did not expect, that a tool designed for HPC like singularity is useful on 32bit. Maybe I can find a 32bit Debian to test it.

DrDaveD replied:
It's probably not very useful, it's just that it's the machine I have handy!  You could leave that testing to me, but what happens on your machine if you leave it as "any"?  debuild appears to figure out what architecture to put in there.

Kunzol replied:
By the way, the semantic of the Architecture line is not the build target.

A value of `all` means that it is an **architecture-independent** package, which can be installed on all architectures. For example a package which only contains bash-scripts. 

This is not true for singularity-container package. It is ether amd64 or i386, depending on the architecture the package is build. It would need two separate Debian packages, one for amd64 and one for i386 (which needs to be built on a i386 or the compiler has to produce i386 binaries.) But this is not the task of "debuild", this must be a configuration in the sourcecode (or mconfig) of singularity.

DrDaveD replied:
The value here though is not `all`, it is `any`.  Does that cause something to break?

Kunzol committed:
Update INSTALL.md

Okay,  looks better.

Co-authored-by: DrDaveD <2129743+DrDaveD@users.noreply.github.com>

Kunzol committed:
mention the place where the debian packages can be found after successful build.

Kunzol committed:
note how to clean up after the build

DrDaveD reviewed (changes requested):
Conflicts need to be resolved by you updating your branch with the latest changes in upstream/master, for example with:
```
git fetch upstream
git checkout master
git merge upstream/master
git push
git checkout - # to switch back to your update-debian-packaging branch
git rebase master
```
You will be asked to resolve conflicts, follow the instructions.  Instead of using rebase `git merge master` should also work but usually rebase is cleaner.  If you don't yet have an upstream remote add it with `git remote add upstream https://github.com/hpcng/singularity.git`.

Very good on the explanation for changes.   At least add "debian" to .gitignore because you want that symlink to stick around.  If the changelog needs to be updated for each tag, then it should be in .github/RELEASE_PROCEDURE.md.  Note that there's another change coming to that file soon due to issue #6019 so you might have another conflict to resolve after changing that.

In summary, the remaining items:
1. Change back the Architecture to "any" unless that causes something to break
2. Add "debian" to .gitignore
3. Add instructions for updating changelog to .github/RELEASE_PROCEDURE.md

Kunzol committed:
added debian link

Kunzol committed:
switched back architecture to any

Kunzol committed:
resolved conflicts

Kunzol committed:
reverted conflict

Kunzol committed:
Update INSTALL.md

Kunzol committed:
updated instruction for debian directory

DrDaveD reviewed (approved):
Awesome.  The only nit I have remaining is that 'dh clean' is no longer needed in the instructions.  Either replace it with 'rm -rf debian' or delete the section on cleaning.  We can get this into 3.8.0.

Kunzol committed:
cleanup extended

Added instruction for removing the copy of the debian directory.
Added a section about the motivation of this.
